### ISSUE TITLE
Allow specifying which Python interpreter to use

### ISSUE THREAD

**Issue by FFY00** [mesonbuild/meson#9671]
There are several use-cases where we might need to specify which Python interpreter we want to use. It would be good if `meson configure` had `--python`/`--python2` arguments to specify the Python interpreter to use.

**Comment by jpakkane** [comment]
Does `/path/to/your/python /path/to/yout/meson configure` not work? And why would you have `--python2` as Meson does not support it (and never has)?

**Comment by FFY00** [comment]
Sorry, I realize I did not make it very clear, I am talking about the interpreter used for the Python modules, not the interpreter that is used to run Meson itself.

**Comment by eli-schwartz** [comment]
This is technically possible today, if you add a "binaries" section to a machine file specifying which python binary should be resolved by program lookup. That would affect both `find_program('python')` and `pymod.find_installation('python')`.

Sadly this is a bit cumbersome to do from the command line, you need to create a machine file just to override the location of one binary lookup.

**Comment by gerion0** [comment]
Relevant: #7078

**Comment by FFY00** [comment]
Specifying a cross file seems to mess things up and make Meson fail to find other programs. In my case, it is failing to find pkgconfig.

Cross file:
```ini
[binaries]
python3 = '/usr/bin/python'
```

Log:
```
+ meson setup --cross-file=/home/anubis/git/mesonpy/tests/packages/purelib-and-platlib/.mesonpy-cross-file.ini --prefix=/usr /home/anubis/git/mesonpy/tests/packages/purelib-and-platlib /tmp/mesonpy-1508cclm/build
The Meson build system
Version: 0.60.3
Source dir: /home/anubis/git/mesonpy/tests/packages/purelib-and-platlib
Build dir: /tmp/mesonpy-1508cclm/build
Build type: cross build
Project name: purelib-and-platlib
Project version: 1.0.0
C compiler for the host machine: ccache cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the host machine: cc ld.bfd 2.36.1
C compiler for the build machine: ccache cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the build machine: cc ld.bfd 2.36.1
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: x86_64
Target machine cpu: x86_64
Found Pkg-config: NO
Run-time dependency python3 found: NO (tried sysconfig)

../../../home/anubis/git/mesonpy/tests/packages/purelib-and-platlib/meson.build:6:0: ERROR: Pkg-config binary for machine MachineChoice.HOST not found. Giving up.

A full log can be found at /tmp/mesonpy-1508cclm/build/meson-logs/meson-log.txt
```

**Comment by FFY00** [comment]
Adding `pkgconfig = 'pkg-config'` to the cross file gets rid of the issue. Is this intended behavior, or should I open a new bug? If this is intended behavior, could we get a mode where the cross files only override default values?

**Comment by eli-schwartz** [comment]
The --cross-file argument indicates to meson that these values are for a cross-compiler configuration. They also activate a cross build.

The --native-file argument indicates to meson that these values are for a native configuration, and does NOT activate a cross build.

Native files and cross files are both variants of "machine files".

Cross builds want you to specify exactly which binaries to use for cross compilation -- the native pkg-config probably won't work by default. ;) So probing for programs is disabled -- not because you specified a cross file, but because you specified a cross *build*. Even then, looking up native binaries e.g. for targets marked as `native: true` will still look up and automatically find various native programs such as pkg-config.

tl;dr when I said above that you can use machine files, I meant `--native-file`.

**Comment by FFY00** [comment]
Thank you for pointing that out, it makes sense. I can confirm using `--native-file` works as expected.
